scripts/asr_language_modelling/ngram_lm/train_kenlm.py

```diff
@@ -0,0 +1,185 @@
+        os.remove(path)
+
+    with Parallel(n_jobs=-2, verbose=10) as parallel:
+        while True:
+            start = current_step * chunk_size
+            end = min((current_step + buffer_size) * chunk_size, dataset_len)
+
+            tokenized_data = parallel(
+                delayed(tokenize_str)(data[start : start + chunk_size], tokenizer, idx_to_token, token_offset)
+                for start in range(start, end, chunk_size)
+            )
+
+            # Write dataset
+            write_dataset(tokenized_data, path)
+
+            current_step += len(tokenized_data)
+
+            print(f"Finished writing {len(tokenized_data)} chunks to {path}. Current chunk index = {current_step}")
+
+            del tokenized_data
+            gc.collect()
+
+            if end >= dataset_len:
+                break
+
+
+def write_dataset(chunks, path):
+    basedir = os.path.dirname(path)
+
+    if not os.path.exists(basedir):
+        os.makedirs(basedir, exist_ok=True)
+
+    with open(path, 'a+', encoding='utf-8') as f:
+        for chunk_idx in tqdm(range(len(chunks)), desc='Chunk ', total=len(chunks), unit=' chunks'):
+            for text in chunks[chunk_idx]:
+                line = ' '.join(text)
+                f.write(f"{line}\n")
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description='Train an N-gram language model with KenLM to be used with beam search decoder of ASR models.'
+    )
+    parser.add_argument(
+        "--train_file",
+        required=True,
+        type=str,
+        help="Path to the training file. It can be a text file or Json manifest.",
+    )
+    parser.add_argument("--nemo_model_file", required=True, type=str)
+    parser.add_argument("--kenlm_model_file", required=True, type=str)
+    parser.add_argument("--ngram_length", required=True, type=int)
+    parser.add_argument("--do_lowercase", action='store_true')
+    parser.add_argument("--kenlm_bin_path", required=True, type=str)
+    args = parser.parse_args()
+
+    """ TOKENIZER SETUP """
+    logging.info(f"Loading nemo model '{args.nemo_model_file}' ...")
+    model = nemo_asr.models.EncDecCTCModelBPE.restore_from(args.nemo_model_file, map_location=torch.device('cpu'))
+
+    encoded_train_file = f"{args.kenlm_model_file}.tmp.txt"
+    """ DATASET SETUP """
+    logging.info(f"Encoding the train file '{args.train_file}' ...")
+    dataset = read_text(args.train_file, lowercase=args.do_lowercase)
+    tokenize_text(
+        dataset,
+        model.tokenizer,
+        path=encoded_train_file,
+        chunk_size=CHUNK_SIZE,
+        buffer_size=CHUNK_BUFFER_SIZE,
+        token_offset=TOKEN_OFFSET,
+    )
+
+    del model
+
+    """ LMPLZ ARGUMENT SETUP """
+    kenlm_args = [
+        os.path.join(args.kenlm_bin_path, 'lmplz'),
+        "-o",
+        f"{args.ngram_length}",
+        "--text",
+        encoded_train_file,
+        "--arpa",
+        f"{args.kenlm_model_file}.tmp.arpa",
+        "--discount_fallback",
```

Comment:
Note that this argument is needed `--discount_fallback` is needed only for subword tokenized text.

Reply:
What does it do exactly?

Reply:
It's used by KenLM to work with subwords instead of assuming character level LM

Reply:
Added